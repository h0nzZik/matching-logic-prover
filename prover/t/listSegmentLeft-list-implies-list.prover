imports "t/definitions/list.prover"
imports "t/definitions/listSegmentLeft.prover"


claim \implies( \and( listSegmentLeft( variable("H") { ArrayIntInt } , variable("X") { Int } , variable("Y") { Int } , variable("F") { SetInt } )
                    , list( variable("H") { ArrayIntInt }, variable("Y") { Int }, variable("G") { SetInt })
                    , \equals( variable("K") { SetInt }, union(variable("F") { SetInt }, variable("G") { SetInt }))
                    , disjoint(variable("F") { SetInt }, variable("G") { SetInt })
                     )
              , \and( list( variable("H") { ArrayIntInt }, variable("X") { Int }, variable("K") { SetInt }) )
              )

/*
Sketch of kt strategy for this proof (only considering recursive case)

lseg(x,y) * list(y) -> list(x)

C[lseg(x, y)] -> list(x) -- defining C == hole * list(y)

lseg(x,y) -> \exists h. h /\ h * list(y) -> list(x) -- by kt-wrap

(x > 0 and lseg(n(x), y))[RHS/lseg][n(x)/x] -> (\exists h. h /\ h * list(y) -> list(x)) -- by lfp

(x > 0 and \exists h. h /\ h * list(y) -> list(n(x))) -> (\exists h. h /\ h * list(y) -> list(x)) -- by applying above substitution

((x > 0 and \exists h. h /\ h * list(y) -> list(n(x))) * list(y)) -> list(x) -- by kt-unwrap

(x > 0 and list(n(x)) * list(y)) -> list(x) -- by kt-collapse
*/

strategy kt ; normalize
            ; ( ( simplify
                ; left-unfold-Nth(0) ;
                  ( ( right-unfold-Nth(0, 0) ; smt)
                  | ( right-unfold-Nth(0, 1) ; smt)
                  )
                )
              )

// strategy search-bound(4)
