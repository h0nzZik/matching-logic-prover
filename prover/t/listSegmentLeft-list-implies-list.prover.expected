<prover>
  <k>
    .
  </k>
  <goals>
    <goal>
      <id>
        root
      </id>
      <active>
        true
      </active>
      <parent>
        .
      </parent>
      <claim>
        \implies ( \and ( listSegmentLeft ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "Y" ) { Int } , variable ( "F" ) { SetInt } , .Patterns ) , list ( variable ( "H" ) { ArrayIntInt } , variable ( "Y" ) { Int } , variable ( "G" ) { SetInt } , .Patterns ) , \equals ( variable ( "K" ) { SetInt } , union ( variable ( "F" ) { SetInt } , variable ( "G" ) { SetInt } , .Patterns ) ) , disjoint ( variable ( "F" ) { SetInt } , variable ( "G" ) { SetInt } , .Patterns ) , .Patterns ) , \exists { .Patterns } \and ( list ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "K" ) { SetInt } , .Patterns ) , .Patterns ) )
      </claim>
      <strategy>
        success
      </strategy>
      <trace>
        .
      </trace>
    </goal>
  </goals>
  <declarations>
    <declaration>
      axiom \forall { variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "F" ) { SetInt } , .Patterns } \iff-lfp ( list ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "F" ) { SetInt } , .Patterns ) , \or ( \exists { .Patterns } \and ( \equals ( variable ( "X" ) { Int } , 0 ) , \equals ( variable ( "F" ) { SetInt } , emptyset ) , .Patterns ) , \exists { variable ( "F1" ) { SetInt } , variable ( "X1" ) { Int } , .Patterns } \and ( list ( variable ( "H" ) { ArrayIntInt } , variable ( "X1" ) { Int } , variable ( "F1" ) { SetInt } , .Patterns ) , gt ( variable ( "X" ) { Int } , 0 , .Patterns ) , \equals ( select ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , .Patterns ) , variable ( "X1" ) { Int } ) , \equals ( variable ( "F" ) { SetInt } , union ( variable ( "F1" ) { SetInt } , singleton ( variable ( "X" ) { Int } , .Patterns ) , .Patterns ) ) , disjoint ( variable ( "F1" ) { SetInt } , singleton ( variable ( "X" ) { Int } , .Patterns ) , .Patterns ) , .Patterns ) , .Patterns ) )
    </declaration> <declaration>
      axiom \forall { variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "Y" ) { Int } , variable ( "F" ) { SetInt } , .Patterns } \iff-lfp ( listSegmentLeft ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "Y" ) { Int } , variable ( "F" ) { SetInt } , .Patterns ) , \or ( \and ( \equals ( variable ( "X" ) { Int } , variable ( "Y" ) { Int } ) , \equals ( variable ( "F" ) { SetInt } , emptyset ) , .Patterns ) , \exists { variable ( "X1" ) { Int } , variable ( "F1" ) { SetInt } , .Patterns } \and ( gt ( variable ( "X" ) { Int } , 0 , .Patterns ) , listSegmentLeft ( variable ( "H" ) { ArrayIntInt } , variable ( "X1" ) { Int } , variable ( "Y" ) { Int } , variable ( "F1" ) { SetInt } , .Patterns ) , \equals ( select ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , .Patterns ) , variable ( "X1" ) { Int } ) , \equals ( variable ( "F" ) { SetInt } , union ( variable ( "F1" ) { SetInt } , singleton ( variable ( "X" ) { Int } , .Patterns ) , .Patterns ) ) , disjoint ( variable ( "F1" ) { SetInt } , singleton ( variable ( "X" ) { Int } , .Patterns ) , .Patterns ) , .Patterns ) , .Patterns ) )
    </declaration> <declaration>
      symbol list ( ArrayIntInt , Int , SetInt , .Sorts ) : Bool
    </declaration> <declaration>
      symbol listSegmentLeft ( ArrayIntInt , Int , Int , SetInt , .Sorts ) : Bool
    </declaration>
  </declarations>
</prover>
